firebase-firestore/src/main/java/com/google/firebase/firestore/remote/AndroidConnectivityMonitor.java

```diff
@@ -0,0 +1,118 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.remote;
+
+import static com.google.firebase.firestore.util.Assert.hardAssert;
+
+import android.annotation.TargetApi;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.net.ConnectivityManager;
+import android.net.Network;
+import android.net.NetworkInfo;
+import android.os.Build;
+import com.google.firebase.firestore.util.Consumer;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Android implementation of ConnectivityMonitor. Parallel implementations exist for N+ and pre-N.
+ *
+ * <p>Implementation note: Most of the code here was shamelessly stolen from
+ * https://github.com/grpc/grpc-java/blob/master/android/src/main/java/io/grpc/android/AndroidChannelBuilder.java
+ */
+public final class AndroidConnectivityMonitor implements ConnectivityMonitor {
+
+  private final Context context;
+  @Nullable private final ConnectivityManager connectivityManager;
+  private final List<Consumer<NetworkStatus>> callbacks = new ArrayList<>();
+
+  public AndroidConnectivityMonitor(Context context) {
+    // This notnull restriction could be eliminated... the pre-N method doesn't
```

Comment:
A possible future refactoring (not for this PR) that would address this would just be to split this into two classes with a shared callback manager helper.

Reply:
FWIW: AndroidChannelBuilder resolves this by (a) checking for null before setting the connectivityManager, and then (b) checking if connectivityManager is null prior to using the N+ variant. (I've preserved that check, though it's not technically necessary here. See line 64.)